### ISSUE TITLE
I want to know how to write training code for Yolov8 in a single machine with multiple GPUs

### ISSUE THREAD

**Issue by baoyuchen** [ultralytics/ultralytics#5742]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I want to know how to write training code for Yolov8 in a single machine with multiple GPUs

**Comment by Kayzwer** [comment]
@baoyuchen just provide "device=0,1,2,3"

**Comment by glenn-jocher** [comment]
@baoyuchen hello,

Training YOLOv8 on multiple GPUs is fairly straightforward, you just need to specify the GPU devices you want to use in your command. You can do this by setting the `device` argument to the IDs of your GPU devices.

For example, if you have four GPUs in your machine with IDs 0, 1, 2, 3, you would specify `device='0,1,2,3'` when calling the train function. This will automatically distribute the training process across the specified GPUs.

You also need to ensure that your batch size is a multiple of the number of GPUs specified to maintain an even distribution of data across the GPUs.

Please note, using multiple GPUs requires a careful adjustment of learning rates, as the effective batch size is the total number of images processed across all GPUs. It is advisable to scale up the learning rate proportionally to the number of GPUs used.

Hope this helps!

Best,
Glenn

**Comment by garygazanga** [comment]
Has anyone successfully trained a model on LOCAL multiple GPUs yet (not DDP)?
I have been scouring the YOLOV8 code but it seems as though (unlike YOLOV5) it only allows you to use Distributed Data Parallel processing (DDP multi-machine) rather than the more straightforward torch.nn.DataParallel() solution.
Specifying GPUs as either "0,1,2,3,5" or [0,1,2,3,5] makes no difference as the 'world_size' becomes greater than 1 and we enter the world of DDP.  There seems to be no way to say "I have multiple GPUs and they are all LOCAL so don't use DDP use DP instead".  YOLOV5 has the line:
                       if cuda and rank == -1 and torch.cuda.device_count() > 1:
                             model = torch.nn.DataParallel(model)
 But my attempts to add something similar to YOLOV8 just results in training stopping at first iteration:
                      File "/app/detector/src/yolo8ultra/ultralytics/utils/loss.py", line 113, in __init__
                      device = next(model.parameters()).device  # get model device
                      StopIteration
 I would really appreciate any help with avoiding the DDP approach (which would involve a big re-write for me).
 Thanks

**Comment by garygazanga** [comment]
I was afraid you might say something like that, but thank you so much for getting back to me so soon!
I have been running the YOLOV5 core code inside my application very successfully for a while now, and converting this to use the core functionality from the YOLOV8 code has proved troublesome.  Perhaps you could clarify something about the DDP mode for me...

myApplication is started with a 'config' parameter and it calls model.train(args) from within.  Because my device arg is a list of GPUs, world_size > 1 and DDP is employed.  However, because I am running the train function from within myApplication, the DDP cmd generated by generate_ddp_command does not contain the 'config' argument and therefore fails to execute under the DDP protocol (it tries to execute myApplication without args, which promptly complains and exits).  So if I wish to adopt the new DDP system - how do I get around this?

I will understand if you regard this question as 'out of scope' - your brain is undoubtedly in high demand!
Thanks for your excellent work +++

**Comment by glenn-jocher** [comment]
Hello @garygazanga,

I understand the difficulty you're encountering with DDP and the integration into your application. In order to get around the problem, and have the DDP command include your 'config' argument, you would need to modify your code to pass the 'config' argument into `train()` function from your application, and subsequently propagate this argument to be included in the DDP command. 

The YOLOv8 codebase is primarily built for command line execution with set programs. In this setup, the training script is normally started directly via the command line, and all arguments are delivered via command line input which DDP automatically propagates.

However, you're using it differently, as a module inside another application and you want to control training from Python, which is causing the DDP command to be generated without the necessary arguments.

Just keep in mind, while `torch.nn.DataParallel` may seem like a more straightforward approach, it doesn't work as efficiently as DDP due to its Python-driven approach for gathering module outputs back onto the main device. Consequently, DDP normally performs better in a multiple GPU scenario.

I'm glad you find the work beneficial and I wish you the utmost luck in your coding endeavors. If you still have further queries, feel free to ask.

Best,
Glenn
